### PR TITLE
[SPARK-13184][SQL] Add a datasource-specific option minPartitions in HadoopFsRelation#options

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?

This pr adds a new option `minPartitions` accepted in `HadoopFsRelation#options`.
This is a datasource-specific value of the minimal splitting number for input data.
## How was this patch tested?

Added unit tests in `FileSourceStrategySuite`.

### CODE REVIEW

**Comment by rxin** [comment]
I was thinking actually just use the same config key name, and override it. Can you read my comment on the jira ticket and let me know if that's clear?

**Comment by maropu** [comment]
oh, sorry. I'll check and fix it.

**Comment by maropu** [comment]
@rxin Could you check this again to satisfy your intention?
Also, I'll make the description up-to-date later.

**Comment by rxin** [comment]
I'm going to mark this for 2.1 given we are too close to 2.0 to modify behaviors like this.

**Comment by maropu** [comment]
Jenkins, retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala [hunk 1]
```diff
@@ -67,9 +67,6 @@ private[sql] abstract class BaseWriterContainer(
   // This is only used on driver side.
   @transient private val jobContext: JobContext = job
 
-  private val speculationEnabled: Boolean =
-    relation.sparkSession.sparkContext.conf.getBoolean("spark.speculation", defaultValue = false)
-
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala]: Deleted because this line not used.

**Comment by maropu** [comment]
@rxin could you give me comments?

**Comment by maropu** [comment]
@gatorsmile Could you check this and give me comments, too?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 2]
```diff
@@ -551,4 +551,10 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
 
   private var extraOptions = new scala.collection.mutable.HashMap[String, String]
 
+  // Returns all option set in the `SparkConf`, the `SQLConf`, and a given data source `options`.
+  // If the same keys exist, they are overridden with ones in the `options`.
+  private def optionsOverriddenWith(options: Map[String, String]): Map[String, String] = {
+    sparkSession.sparkContext.conf.getAllAsMap ++ sparkSession.conf.getAll ++ options
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: Here, we just cover the code path for DataFrameReader APIs.

**Comment by gatorsmile** [comment]
To fundamentally resolve this issue, we need to consider all the read/write paths. A lot of works to do for achieving it. Do you want to continue the work or just close it?

**Comment by maropu** [comment]
Ah, ok. I'll check again.

**Comment by HyukjinKwon** [comment]
@maropu, I just wonder if it is in progress in any way.

**Comment by maropu** [comment]
No, so I'll close this for now and we move this discussion to jira.

**Pull request closed without merging**